sled-agent/src/services.rs

```diff
@@ -327,6 +333,152 @@ impl ServiceManager {
+            &self.log,
+            &self.vnic_allocator,
+            &zone_name,
+            // unique_name=
+            None,
+            // dataset=
+            &[],
+            &devices,
+            // opte_ports=
+            vec![],
+            // physical_nic=
+            physical_nic,
+        )
+        .await?;
+
+        let running_zone = RunningZone::boot(installed_zone).await?;
+
+        for addr in addresses {
+            info!(self.log, "Ensuring address {} exists", addr.to_string());
+            let addr_request =
+                AddressRequest::new_static(IpAddr::V6(*addr), None);
+            running_zone.ensure_address(addr_request).await?;
+            info!(
+                self.log,
+                "Ensuring address {} exists - OK",
+                addr.to_string()
+            );
+        }
+
+        info!(self.log, "GZ addresses: {:#?}", gz_addresses);
+        for &addr in gz_addresses {
+            info!(self.log, "Ensuring GZ address {} exists", addr.to_string());
+
+            let addr_name = zone_name.replace(&['-', '_'][..], "");
+            Zones::ensure_has_global_zone_v6_address(
+                self.underlay_vnic.clone(),
+                addr,
+                &addr_name,
+            )
+            .map_err(|err| Error::GzAddress {
+                message: format!(
+                    "adding address on behalf of service zone '{}'",
+                    zone_name
+                ),
+                err,
+            })?;
+
+            // If this address is in a new ipv6 prefix, notify maghemite so
+            // it can advertise it to other sleds.
+            self.advertise_prefix_of_address(addr).await;
+        }
+
+        let gateway = if !gz_addresses.is_empty() {
+            // If this service supplies its own GZ address, add a route.
+            //
+            // This is currently being used for the DNS service.
+            //
+            // TODO: consider limiting the number of GZ addresses which
+            // can be supplied - now that we're actively using it, we
+            // aren't really handling the "many GZ addresses" case, and it
+            // doesn't seem necessary now.
+            gz_addresses[0]
+        } else {
+            self.underlay_address
+        };
+
+        running_zone.add_default_route(gateway).await.map_err(|err| {
+            Error::ZoneCommand { intent: "Adding Route".to_string(), err }
+        })?;
+
+        Ok(running_zone)
+    }
+
+    // Check the services intended to run in the zone to determine whether any
+    // physical devices need to be mapped into the zone when it is created.
+    fn devices_needed(
+        &self,
+        req: &ServiceZoneRequest,
+    ) -> Result<Vec<String>, Error> {
+        // When running on a real sidecar, we need the /dev/tofino device
+        // to talk to the tofino ASIC.
+        // TODO: this really needs to be checked periodically rather
```

Comment:
There would need to be some OS work, but I don't think it's insurmountable.  We'd need to be able to trigger reevaluation of the match rules in the configuration, probably in response to devfsadm or other sysevent activity, and create any missing nodes in the zone `/dev` directory.

Reply:
As a short-term workaround, would it work to reboot the Dendrite zone if a new `tofino` device appears? We could monitor for this in the GZ, and react by rebooting the zone.

Reply:
When built for `tofino-asic`, `dpd` will currently crash if `/dev/tofino` isn't present.  I can change that easily enough.  With that change, then rebooting when the tofino device (dis)appears should work.

Then the question becomes: if we're not keying off the presence of the device,  when do we launch a `switch` zone?  Do we really want that to be running on every gimlet?

Reply:
> When built for `tofino-asic`, `dpd` will currently crash if `/dev/tofino` isn't present. I can change that easily enough. With that change, then rebooting when the tofino device (dis)appears should work.

Sure, though would this change even be necessary? Presumably, the GZ could wait until such a device appears before launching the switch zone.

> Then the question becomes: if we're not keying off the presence of the device, when do we launch a `switch` zone? Do we really want that to be running on every gimlet?

I do not think so - I think we want the `switch` zone software to exist on all gimlets (in case one is moved to the scrimlet location), but I think we'd prefer to refrain from launching the zone until we have confirmation this is a scrimlet device.

In https://github.com/oxidecomputer/omicron/pull/1510 I added a hard-coded configuration to make this distinction (if you wanted to test this now), but I do think the decision of "scrimlet vs gimlet" should be dynamically detectable, as discussed in https://github.com/orgs/oxidecomputer/projects/19/views/1